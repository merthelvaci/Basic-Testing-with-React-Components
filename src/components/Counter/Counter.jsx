import React, { useCallback, useState } from 'react';

const Counter = () => {
	const [counter, setCounter] = useState(0);

	const incrementCounterHandler = useCallback(() => {
		setCounter((prevState) => prevState + 1);
	}, [counter]);

	const decrementCounterHandler = useCallback(() => {
		setTimeout(() => {
			setCounter((prevState) => prevState - 1);
		}, 500);
	}, [counter]);

	return (
		<div className="counter-wrapper">
			<button data-testid="incBtn" onClick={incrementCounterHandler}>
				Increment
			</button>
			<button data-testid="decBtn" onClick={decrementCounterHandler}>
				Decrement
			</button>
			<span data-testid="counterText">Counter: {counter}</span>
		</div>
	);
};

export default Counter;
