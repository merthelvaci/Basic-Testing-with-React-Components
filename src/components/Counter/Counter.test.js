import Counter from '.';

let asFrag;
let byTestId;
let byText;
beforeEach(() => {
	const { asFragment, getByTestId, getByText } = render(<Counter />);
	asFrag = asFragment;
	byTestId = getByTestId;
	byText = getByText;
});

afterEach(cleanup);

describe('<Counter /> component tests', () => {
	it('should render <Counter /> component correctly', () => {
		expect(asFrag()).toMatchSnapshot();
	});

	it('should render counter as 0 at startup', () => {
		expect(byTestId('counterText')).toHaveTextContent('0');
	});

	it('should increment counter upon clicking "Increment" button', () => {
		const incButtonEl = byTestId('incBtn');
		fireEvent.click(incButtonEl);
		expect(byTestId('counterText')).toHaveTextContent('1');
	});

	it('should dcrement counter asynchronously upon clicking "Decrement" button', async () => {
		const decButtonEl = byTestId('decBtn');
		fireEvent.click(decButtonEl);

		const counterSpan = await waitForElement(() => byText('Counter: -1'));
		expect(counterSpan).toHaveTextContent('-1');
	});
});
