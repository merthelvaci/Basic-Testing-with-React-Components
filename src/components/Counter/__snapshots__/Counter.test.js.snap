// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> component tests should render <Counter /> component correctly 1`] = `
<DocumentFragment>
  <div
    class="counter-wrapper"
  >
    <button
      data-testid="incBtn"
    >
      Increment
    </button>
    <button
      data-testid="decBtn"
    >
      Decrement
    </button>
    <span
      data-testid="counterText"
    >
      Counter: 0
    </span>
  </div>
</DocumentFragment>
`;
